File: python/src/acedatacloud_x402/handler.py
"""SDK adapter: turn X402 credentials into a ``payment_handler`` for the SDK.

Compatible with both the sync and async :class:`AceDataCloud` clients in
the ``acedatacloud`` package. The returned callable matches the
``SyncPaymentHandler`` signature; the async client will happily await a
sync result too.
"""

from __future__ import annotations

import base64
import json
from collections.abc import Callable, Sequence
from typing import Any

from .signing.evm import EVMAccountSigner, sign_evm_payment, sign_evm_upto_payment
from .signing.solana import SolanaKeypairSigner, sign_solana_payment
from .types import Network, PaymentRequirement, X402PaymentEnvelope

PaymentHandlerResult = dict[str, dict[str, str]]
NETWORK_ALIASES = {
    "base": "eip155:8453",
    "skale": "eip155:1187947933",
    "solana": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
}


def _encode_payment_header(envelope: X402PaymentEnvelope) -> str:
    serialized = json.dumps(envelope, separators=(",", ":"))
    return base64.b64encode(serialized.encode("utf-8")).decode("ascii")


def _select_requirement(
    accepts: Sequence[PaymentRequirement],
    network: str,
    *,
    prefer_scheme: str | None = None,
) -> PaymentRequirement | None:
    """Pick the best matching accept entry for ``network``.

    If ``prefer_scheme`` is provided and a match exists, that variant wins.
    Otherwise the first network match is returned (preserving server ordering).
    """
    matches = [req for req in accepts if req.get("network") == network]
    if not matches:
        return None
    if prefer_scheme:
        for req in matches:
            if req.get("scheme") == prefer_scheme:
                return req
    return matches[0]


def create_x402_payment_handler(
    *,
    network: Network,
    evm_signer: EVMAccountSigner | None = None,
    solana_signer: SolanaKeypairSigner | None = None,
    rpc_url: str | None = None,
    prefer_scheme: str | None = None,
) -> Callable[[dict[str, Any]], PaymentHandlerResult]:
    """Build a ``payment_handler`` function the SDK can invoke on 402.

    Args:
        network: Target payment network (``"solana"``, ``"base"``, or ``"skale"``).
        evm_signer: Required for ``base`` / ``skale``.
        solana_signer: Required for ``solana``.
        rpc_url: Optional Solana RPC override.
        prefer_scheme: When the server offers multiple schemes for the chosen
            network, prefer this one (e.g. ``"upto"`` for chat APIs with
            metered billing, ``"exact"`` for fixed-price endpoints). Falls
            back to the first entry the server returns.

    Returns:
        A callable ``(ctx) -> {"headers": {"X-Payment": "..."}}`` suitable
        for :class:`acedatacloud.AceDataCloud` /
        :class:`acedatacloud.AsyncAceDataCloud`.
    """
    canonical_network = NETWORK_ALIASES.get(network, network)
    is_solana = canonical_network.startswith("solana:")
    is_evm = canonical_network.startswith("eip155:")
    if is_solana and solana_signer is None:
        raise ValueError('solana_signer is required when network="solana"')
    if is_evm and evm_signer is None:
        raise ValueError(f'evm_signer is required when network="{network}"')

    def handler(ctx: dict[str, Any]) -> PaymentHandlerResult:
        accepts: Sequence[PaymentRequirement] = ctx.get("accepts") or []
        requirement = _select_requirement(accepts, canonical_network, prefer_scheme=prefer_scheme)
        if requirement is None:
            available = ", ".join(str(r.get("network")) for r in accepts) or "<none>"
            raise RuntimeError(
                f'X402: no payment requirement for network "{canonical_network}". '
                f"Available: {available}"
            )

        if is_solana:
            assert solana_signer is not None  # guarded above
            envelope = sign_solana_payment(requirement, solana_signer, rpc_url=rpc_url)
        else:
            assert evm_signer is not None  # guarded above
            if requirement.get("scheme") == "upto":
                envelope = sign_evm_upto_payment(requirement, evm_signer)
            else:
                envelope = sign_evm_payment(requirement, evm_signer)

        return {"headers": {"PAYMENT-SIGNATURE": _encode_payment_header(envelope)}}

    return handler

File: go/handler.go
package x402

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
)

// HandlerOptions configures a Handler. At least one signer must be set
// for the selected network.
type HandlerOptions struct {
	Network      Network
	EVMSigner    *EVMSigner
	SolanaSigner *SolanaSigner
	// RPCURL optionally overrides the Solana RPC endpoint. Ignored for EVM.
	RPCURL string
	// HTTPClient optionally overrides the http.Client used for Solana RPC.
	HTTPClient *http.Client
}

// Handler produces ``X-Payment`` headers for a given 402 accepts list.
//
// It is intentionally self-contained — it does NOT depend on the
// AceDataCloud SDK's ``PaymentHandler`` interface. Plug it into the SDK
// with a trivial 4-line adapter; see the README.
type Handler struct {
	opts HandlerOptions
}

// NewHandler constructs a Handler and validates its options.
func NewHandler(opts HandlerOptions) (*Handler, error) {
	switch opts.Network {
	case NetworkBase, NetworkSKALE:
		if opts.EVMSigner == nil {
			return nil, fmt.Errorf(`x402: EVMSigner is required when network=%q`, opts.Network)
		}
	case NetworkSolana:
		if opts.SolanaSigner == nil {
			return nil, errors.New(`x402: SolanaSigner is required when network="solana"`)
		}
	case "":
		return nil, errors.New("x402: Network is required")
	default:
		return nil, fmt.Errorf("x402: unsupported network %q", opts.Network)
	}
	return &Handler{opts: opts}, nil
}

// Sign picks the matching requirement from ``accepts``, signs the
// payment (and for Solana submits it on-chain), and returns the
// ``X-Payment`` HTTP header value.
func (h *Handler) Sign(ctx context.Context, accepts []PaymentRequirement) (string, error) {
	req, err := h.selectRequirement(accepts)
	if err != nil {
		return "", err
	}
	var envelope *X402PaymentEnvelope
	switch h.opts.Network {
	case NetworkSolana:
		envelope, err = SignSolanaPayment(ctx, req, h.opts.SolanaSigner, h.opts.RPCURL, h.opts.HTTPClient)
	default:
		envelope, err = SignEVMPayment(req, h.opts.EVMSigner)
	}
	if err != nil {
		return "", err
	}
	return EncodePaymentHeader(envelope)
}

// Headers is a convenience that wraps Sign and returns a
// ``{"X-Payment": "<value>"}`` map, ready to merge into HTTP headers.
func (h *Handler) Headers(ctx context.Context, accepts []PaymentRequirement) (map[string]string, error) {
	value, err := h.Sign(ctx, accepts)
	if err != nil {
		return nil, err
	}
	return map[string]string{"X-Payment": value}, nil
}

// selectRequirement returns the first accepts entry whose network matches
// the configured handler network.
func (h *Handler) selectRequirement(accepts []PaymentRequirement) (PaymentRequirement, error) {
	for _, a := range accepts {
		if a.Network == string(h.opts.Network) {
			return a, nil
		}
	}
	var available string
	for i, a := range accepts {
		if i > 0 {
			available += ", "
		}
		available += a.Network
	}
	if available == "" {
		available = "<none>"
	}
	return PaymentRequirement{}, fmt.Errorf("x402: no payment requirement for network %q. Available: %s", h.opts.Network, available)
}

// EncodePaymentHeader serializes and base64-encodes an envelope.
func EncodePaymentHeader(envelope *X402PaymentEnvelope) (string, error) {
	raw, err := json.Marshal(envelope)
	if err != nil {
		return "", fmt.Errorf("x402: marshal envelope: %w", err)
	}
	return base64.StdEncoding.EncodeToString(raw), nil
}
